gen/service.go

```diff
@@ -0,0 +1,225 @@
+		Name:   goCase(f.Name) + "Args",
+		Fields: compile.FieldGroup(f.ArgsSpec),
+	}
+	if err := argsGen.Generate(g); err != nil {
+		return wrapGenerateError(fmt.Sprintf("%s.%s", s.Name, f.Name), err)
+	}
+
+	resultFields := make(compile.FieldGroup, 0)
+	if result := f.ResultSpec; result != nil {
+		if result.ReturnType != nil {
+			resultFields = append(resultFields, &compile.FieldSpec{
+				ID:   0,
+				Name: "success",
+				Type: result.ReturnType,
+			})
+		}
+
+		resultFields = append(resultFields, result.Exceptions...)
+	}
+	resultGen := fieldGroupGenerator{
+		Name:   goCase(f.Name) + "Result",
+		Fields: resultFields,
+	}
+	if err := resultGen.Generate(g); err != nil {
+		return wrapGenerateError(fmt.Sprintf("%s.%s", s.Name, f.Name), err)
+	}
+
+	if err := functionHelper(g, f); err != nil {
+		return wrapGenerateError(fmt.Sprintf("%s.%s", s.Name, f.Name), err)
+	}
+
+	// TODO(abg): If we receive unknown exceptions over the wire, we need to
+	// throw a generic error.
+	return nil
+}
+
+func functionHelper(g Generator, f *compile.FunctionSpec) error {
+	return g.DeclareFromTemplate(
+		`
+		<$name := goCase .Name>
+
+		var <$name> = struct{
+			IsException func(error) bool
+
+			Args func(
+				<range .ArgsSpec>
+					<if .Required>
+						<.Name> <typeReference .Type>,
+					<else>
+						<.Name> <typeReferencePtr .Type>,
+					<end>
+				<end>
+			) *<$name>Args
+
+			<if .HasReturn>
+				WrapResponse func(<typeReferencePtr .ResultSpec.ReturnType>, error) (*<$name>Result, error)
+				UnwrapResponse func(*<$name>Result) (<typeReferencePtr .ResultSpec.ReturnType>, error)
+			<else>
+				WrapResponse func(error) (*<$name>Result, error)
+				UnwrapResponse func(*<$name>Result) error
+			<end>
+		}{}
+
+		func init() {
+			<$name>.IsException = func(err error) bool {
+				switch err.(type) {
+				<if .ResultSpec>
+					<range .ResultSpec.Exceptions>
+					case <typeReferencePtr .Type>:
+						return true
+					<end>
+				<end>
+				default:
+					return false
+				}
+			}
+
+			<$name>.Args = func(
+				<range .ArgsSpec>
+					<if .Required>
+						<.Name> <typeReference .Type>,
+					<else>
+						<.Name> <typeReferencePtr .Type>,
+					<end>
+				<end>
+			) *<$name>Args {
+				return &<$name>Args{
+				<range .ArgsSpec>
+					<if .Required>
+						<goCase .Name>: <.Name>,
+					<else>
+						<goCase .Name>: <.Name>,
+					<end>
+				<end>
+				}
+			}
+
+			<$name>.WrapResponse =
+			<if .HasReturn>
+				func(success <typeReferencePtr .ResultSpec.ReturnType>, err error) (*<$name>Result, error) {
+					if err == nil {
+						return &<$name>Result{Success: success}, nil
+					}
+			<else>
+				func(err error) (*<$name>Result, error) {
+					if err == nil {
+						return &<$name>Result{}, nil
+					}
+			<end>
+					<if .ResultSpec>
+						<if .ResultSpec.Exceptions>
+							switch e := err.(type) {
+								<range .ResultSpec.Exceptions>
+								case <typeReferencePtr .Type>:
+									return &<$name>Result{<goCase .Name>: e}, nil
```

Comment:
some people hit the `err != nil` but it's actually a `nil` error with a type,
https://github.com/uber/tchannel-go/issues/233

It might be worth checking in here
```
if e == nil {
  // do something
}
```
here's the fix in thrift-gen,
https://github.com/uber/tchannel-go/pull/240/files